### PR TITLE
storageautoscaler: scale only when more than threshold minus 10% percent of cluster is used

### PR DESCRIPTION
**Pull request by parth-gr**
add a another check for scaling to scale only when more than threshold minus 10% of cluster is full

in some cases ceph takes time to balance the osd,
so its important to calculate the total cluster size for a device class

### CODE REVIEW

File: controllers/storageautoscaler/scraper.go [hunk 1]
```diff
@@ -107,9 +112,11 @@ func filterObjectsForScaling(storageAutoScalerList *ocsv1.StorageAutoScalerList,
 			err := fmt.Errorf("osd usage is nil for device class, device class: %s", storageAutoScaler.Spec.DeviceClass)
 			log.Error(err, "device class not found in sync map")
 		} else {
-			deviceClassUsage := usage.(float64)
-			if (deviceClassUsage * 100) >= float64(deviceClassThreshold) {
-				objectList = append(objectList, types.NamespacedName{Namespace: storageAutoScaler.Namespace, Name: storageAutoScaler.Name})
+			deviceClassUsage := usage.(OsdUsage)
+			if deviceClassUsage.TotalUsage*100 >= 50 {
```
Comment (travisn) [inline controllers/storageautoscaler/scraper.go:116]: Let's make this more dependent on the configured threshold. After some quick calculations I would suggest `deviceClassThreshold - 10%`. By default, this would be a 60% threshold, which gives some room for OSDs to have different filled ratios, but also not too low of a threshold. If the threshold is too low, it may not work as well when we are horizontally scaling and adding 3 OSDs (instead of doubling OSD sizes in vertical expansion). Proportionally, we won't be adding as much per horizontal expansion as the cluster gets bigger.

Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go:116]: Correct I did some calculations too,

if we are scaling horizontally after 8tb osds,

Total capacity we increase like, 
24, 48, 72, 96......, 216, 240, 268...

248 ---> 268 => 10% increase only....

Which will fail the above algorithm if someone if going above 240Tb.....

And the max supported we can go to is 768Tb

Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go:116]: (768-744)/744 = 3.2% approx

So we can have a limit saying atleast (threshod-3%) is full??

Comment (travisn) [reply controllers/storageautoscaler/scraper.go:116]: The OSD filled percentage will always have some variability. 3% is too small. As discussed offline, horizontal scaling a small percentage such as 3% is really the issue. I'd suggest:
1. Use `threshold - 10%` for this secondary decision point (in this PR)
2. Horizontal scaling should increase the storage capacity at least 10% (in a separate PR). Instead of adding 3 OSDs, we will need to add some multiple of 3 OSDs to make it expand at least 10%. This means larger clusters will expand more proportionally.

Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go]: We can open a separate discussion of second point.

File: controllers/storageautoscaler/scraper.go [hunk 2]
```diff
@@ -107,9 +112,11 @@ func filterObjectsForScaling(storageAutoScalerList *ocsv1.StorageAutoScalerList,
 			err := fmt.Errorf("osd usage is nil for device class, device class: %s", storageAutoScaler.Spec.DeviceClass)
 			log.Error(err, "device class not found in sync map")
 		} else {
-			deviceClassUsage := usage.(float64)
-			if (deviceClassUsage * 100) >= float64(deviceClassThreshold) {
-				objectList = append(objectList, types.NamespacedName{Namespace: storageAutoScaler.Namespace, Name: storageAutoScaler.Name})
+			deviceClassUsage := usage.(OsdUsage)
+			if deviceClassUsage.TotalUsage*100 >= math.Max(float64(deviceClassThreshold)-10, 0) {
+				if (deviceClassUsage.HighestUsage * 100) >= float64(deviceClassThreshold) {
```
Comment (travisn) [inline controllers/storageautoscaler/scraper.go:117]: nit: It may be helpful to log if the highest usage OSD is above the threshold, but we are skipping the expansion. So you may want to switch the order of these checks, and log in that case?

File: controllers/storageautoscaler/scraper.go (hunk 1 above)
Comment (alfonsomthd) [reply controllers/storageautoscaler/scraper.go]: @parth-gr @travisn are these new calculations affecting the way the UI calculates all the discrete values for the cluster capacity?

Comment (travisn) [reply controllers/storageautoscaler/scraper.go]: @alfonsomthd I expect those discrete values will all still be valid. 
@parth-gr One consideration is that when we hit the max capacity, we only scale to that capacity (and not horizontally scale too many OSDs past that). Is that already in the calculation?

Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go]: If we are scaling horizontally, we add 3 osds, and if we cant add 3 more osd we do not scale, https://github.com/red-hat-storage/ocs-operator/blob/main/controllers/storageautoscaler/storageautoscaler_controller.go#L165

>I expect those discrete values will all still be valid.

It will change as we will be scaling to specific numbers

Comment (travisn) [reply controllers/storageautoscaler/scraper.go]: > If we are scaling horizontally, we add 3 osds, and if we cant add 3 more osd we do not scale,

Understood, though let me give an example just to be clear... Let's say we have a cluster with max capacity 400TiB
and current capacity is 360TiB...
When we reach the 70% threshold:
   - Expansion of 3 OSDs (8TiB each) would take us to 384TiB capacity. 
   - This is only a 6.7% expansion and we want a minimum of 10% expansion if possible. 
   - So with the new proposal we would consider adding 6 OSDs to get at least 10% expansion. But that would give us 408TiB, which is over the 400TiB capacity, which is too much. 
   - So we just fall back to 3 OSDs expansion in this case.

Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go]: Agree, we can do that, but that also means, 10% is the min. increase?

as we have 3 osd with 8 tb and we add three more, we would be increasing 50% of the cluster.

Comment (travisn) [reply controllers/storageautoscaler/scraper.go]: Yes, let's make 10% the min increase. For large clusters, horizontal scale of 3 OSDs is not enough of a jump.

Comment (alfonsomthd) [reply controllers/storageautoscaler/scraper.go]: @parth-gr Can you update in the design doc the new algorithm? We need to update the calculations also on UI.

Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go]: @alfonsomthd as planned we will be doing it as part of 4.20

File: controllers/storageautoscaler/scraper.go [hunk 3]
```diff
@@ -71,10 +77,10 @@ func (r *StorageAutoscalerScraper) enqueueGenericEventsFromMetrics(ctx context.C
 // update the sync map with the highest value of osd usage per device class
 func updateSyncMap(metrics model.Vector, syncMap *sync.Map) {
 	// create a map of device class to highest osd usage
-	deviceClassUsage := make(map[string]float64)
+	deviceClassUsage := make(map[string]OsdUsage)
 	for _, osd := range metrics {
 		deviceClass := string(osd.Metric["device_class"])
-		deviceClassUsage[deviceClass] = math.Max(deviceClassUsage[deviceClass], float64(osd.Value))
+		deviceClassUsage[deviceClass] = OsdUsage{HighestUsage: math.Max(deviceClassUsage[deviceClass].HighestUsage, float64(osd.Value)), TotalUsage: deviceClassUsage[deviceClass].TotalUsage + float64(osd.Value), TotalOsd: deviceClassUsage[deviceClass].TotalOsd + 1}
```
[the remark this answers is not in the document]
Comment (travisn) [reply controllers/storageautoscaler/scraper.go:83]: Or even better, have a helper method create this struct and unit test it

File: controllers/storageautoscaler/scraper.go [hunk 4]
```diff
@@ -107,16 +113,23 @@ func filterObjectsForScaling(storageAutoScalerList *ocsv1.StorageAutoScalerList,
...
+				}
+				log.Info("highest osd usage is greater than the threshold, but total cluster usage is less than the (threshold-10)%, skipping scaling", "device class", deviceClass, "highest osd usage", deviceClassUsage.HighestUsage*100, "total osd usage", deviceClassUsage.TotalUsage*100)
 			}
 		}
 	}
 
 	return objectList
 }
 
+func totalUsageGreaterThanThresholdMinusTenPercent(deviceClassUsage OsdUsage, deviceClassThreshold int) bool {
```
Comment (travisn) [inline controllers/storageautoscaler/scraper.go:129]: nit: don't bake the 10% threshold into the method name
```suggestion
func totalUsageGreaterThanAdjustedThreshold(deviceClassUsage OsdUsage, deviceClassThreshold int) bool {
```

File: controllers/storageautoscaler/scraper.go [hunk 5]
```diff
@@ -107,16 +113,23 @@ func filterObjectsForScaling(storageAutoScalerList *ocsv1.StorageAutoScalerList,
...
+	return (deviceClassUsage.TotalUsage/deviceClassUsage.TotalOsd)*100 >= math.Max(float64(deviceClassThreshold)-1, 0)
```
Comment (travisn) [inline controllers/storageautoscaler/scraper.go:130]: Check if `TotalOsd` is 0, otherwise you could have divide by 0.

Comment (travisn) [reply controllers/storageautoscaler/scraper.go:130]: Wait, we need a different computation. My comment [here](https://github.com/red-hat-storage/ocs-operator/pull/3200#discussion_r2080388652) must not be clear. The 10% difference to check is that if we expand horizontally, that we are expanding at least 10% of the capacity. If adding 3 OSDs will result in less than 10% expansion, then we need to have some multiple of 3 OSDs until it is at least 10%.

An exception to this rule would be if adding more than 3 OSDs will exceed the total autoscale storage limit, then we can just expand 3.

File: controllers/storageautoscaler/scraper.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go:83]: Its a straight forward input, i dont see a value to unit test it

File: controllers/storageautoscaler/scraper.go (hunk 5 above)
Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go:130]: we will do that process in the next release, currently sticking to add only 3 osds at a time

**Comment by parth-gr** [comment]
Testing looks good,

for testing I did (threshod-1) (8-1=7) as the cluster was less filled, but worked fine,

After first resize, it didnt did a second resize, and the log was like 

```
```
{"level":"info","ts":"2025-05-14T14:21:34Z","logger":"scraper.StorageAutoscalerScraper","msg":"highest osd usage is greater than the threshold, but total cluster usage is less than the (threshold-10)%, skipping scaling","device class":"ssd","highest osd usage":9.021980285644531,"total osd usage":4.669428507486979}
```

```

File: controllers/storageautoscaler/scraper_test.go [hunk 6]
```diff
@@ -156,6 +164,27 @@ func TestUpdateSyncMap(t *testing.T) {
...
+		deviceClassUsage := OsdUsage{
+			TotalUsage:   1.80,
+			TotalOsd:     3,
+			HighestUsage: 0.60,
+		}
+		result := totalUsageGreaterThanAdjustedThreshold(deviceClassUsage, 50)
+		assert.True(t, result)
+	})
+	t.Run("after resize test total usage less than threshold minus 10 percent", func(t *testing.T) {
+		deviceClassUsage := OsdUsage{
```
Comment (travisn) [inline controllers/storageautoscaler/scraper_test.go:182]: How about adding a comment here to explain more? I had to think about this too long. :) 
For example, that the average is 32% so it's less than the 40% adjusted threshold.

File: controllers/storageautoscaler/scraper.go (hunk 5 above)
Comment (travisn) [reply controllers/storageautoscaler/scraper.go]: Oh right, that will be a future PR. The calculation for this purpose looks good now.

File: controllers/storageautoscaler/scraper_test.go [hunk 7]
```diff
@@ -156,6 +164,27 @@ func TestUpdateSyncMap(t *testing.T) {
...
+			TotalUsage:   1.90,
+			TotalOsd:     6,
+			HighestUsage: 0.55,
+		}
```
Comment (travisn) [inline controllers/storageautoscaler/scraper_test.go:186]: How about adding a test case that is barely over 40% adjusted threshold? Boundary checking is helpful for computations like this.

File: controllers/storageautoscaler/scraper.go [hunk 8]
```diff
@@ -107,16 +117,30 @@ func filterObjectsForScaling(storageAutoScalerList *ocsv1.StorageAutoScalerList,
 			err := fmt.Errorf("osd usage is nil for device class, device class: %s", storageAutoScaler.Spec.DeviceClass)
 			log.Error(err, "device class not found in sync map")
 		} else {
-			deviceClassUsage := usage.(float64)
-			if (deviceClassUsage * 100) >= float64(deviceClassThreshold) {
-				objectList = append(objectList, types.NamespacedName{Namespace: storageAutoScaler.Namespace, Name: storageAutoScaler.Name})
+			deviceClassUsage := usage.(OsdUsage)
+			if (deviceClassUsage.HighestUsage * 100) >= float64(deviceClassThreshold) {
+				if deviceClassUsage.TotalOsd == 0 {
```
Comment (travisn) [inline controllers/storageautoscaler/scraper.go:122]: It's best to check for 0 in the same place of the potential divide by zero method. So I'd suggest to move this check into the `totalUsageGreaterThanAdjustedThreshold()` method, log the message, and then just return false in that case. Then you can unit test that case too.

**Review by travisn — CHANGES_REQUESTED** [review]

File: controllers/storageautoscaler/scraper.go (hunk 8 above)
Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go:122]: we have a another log message at line 132, `highest osd usage is greater than the threshold, but t` 

where we again need to divide it by osd_count, so thats why had a big if with it,

Not sure how can we move that log, but will think again

Comment (travisn) [reply controllers/storageautoscaler/scraper.go:122]: How about add a helper method for `averageOsdUsage()`, and then it can check for the divide by 0. If number of OSDs is 0, it could just log the error and return 0. That helper can be called from both the log message and `totalUsageGreaterThanAdjustedThreshold()`

**Review by travisn — COMMENTED** [review]
LGTM, perhaps just another small unit test

**Review by travisn — COMMENTED** [review]
/lgtm

File: controllers/storageautoscaler/scraper.go [hunk 9]
```diff
@@ -99,24 +109,47 @@ func filterObjectsForScaling(storageAutoScalerList *ocsv1.StorageAutoScalerList,
...
-			}
+		if checkIfScalingRequired(syncMap, deviceClassThreshold, log, deviceClass) {
+			objectList = append(objectList, types.NamespacedName{Namespace: storageAutoScaler.Namespace, Name: storageAutoScaler.Name})
 		}
 	}
 
 	return objectList
 }
 
+func checkIfScalingRequired(syncMap *sync.Map, deviceClassThreshold int, log logr.Logger, deviceClass string) bool {
```
Comment (alfonsomthd) [inline controllers/storageautoscaler/scraper.go:120]: Just a simpler name suggestion:
```suggestion
func isScalingRequired(syncMap *sync.Map, deviceClassThreshold int, log logr.Logger, deviceClass string) bool {
```

Comment (parth-gr) [reply controllers/storageautoscaler/scraper.go]: Done

Comment (alfonsomthd) [reply controllers/storageautoscaler/scraper.go]: Tagged!
